design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,178 @@
+This proposal discusses how to fix this, while ensuring that existing VMs keep their current ID.
+
+## Goals
+* Generate and persist universally-unique Firmware.UUID for new VMs
+* Keep the Firmware.UUID of pre-existing VMs
+
+## Non Goals
+
+
+## Definition of Users
+End Users: Individuals or organizations running VMs and VMIs on KubeVirt who require consistent firmware UUIDs for their applications.
+
+## User Stories
+As an end-user, I expect my VMI to maintain its identity across restarts.
+
+## Repos
+Kubevirt/kubevirt
+
+# Design
+
+This section provides a range of alternative solutions for implementing a persistent,
+universally unique firmware UUID for Virtual Machines Instances (VMIs) in KubeVirt.
+
+### 1. Persist UUID in the VM's Spec Field
+
+**Description:** The UUID is generated when the VMI is being created and is then saved to the VM's spec field.
+
+**Pros:**
+- Straightforward implementation.
+- Avoids the need to introduce new API fields.
+- Fully backward compatible.
+
+**Cons:**
+- Potential misuse of the spec field, which ideally should only reflect the user’s intended configuration.
+- The majority of users may not consider the UUID part of the desired VM state, making this field irrelevant from their perspective.
+- Can lead to longer and more complex VM definitions, impacting readability and management.
+
+---
+
+### 2. Generate a Default UUID via Webhook and Set it to the Spec Field
+
+**Description:** A webhook generates a UUID if the user does not provide one. This UUID is then set in the VM's spec field.
+
+**Pros:**
+- Retains backward compatibility and avoids introducing new API fields.
+
+**Cons:**
+- Same concerns as above about the spec field.
+- Using webhooks to assign defaults can degrade performance, especially with high volumes of VMs, as `virt-api` may become a bottleneck.
+- May compromise scalability, as generating a UUID via webhook adds overhead when multiple VMs start simultaneously.
+
+---
+
+### 3. Create a New Firmware UUID Field in VM Status
+
+**Description:** A new field is added under VM status to store the generated firmware UUID upon first VM boot.
+
+**Pros:**
+- Keeps the spec clean, avoiding unnecessary fields in the user-defined configuration.
+- Provides a clear separation between user configuration (spec) and system-generated information (status).
+
+**Cons:**
+- Introduces a new API field, which could increase API surface area and requires long-term support.
+- New fields may impact system performance and resource load.
+- Adds to API complexity and may reduce readability.
+
+*Note:* To minimize this con, consider setting the UUID as a condition under VM status instead of introducing a separate field.
+
+---
+
+### 4. Introduce a Breaking Change to Ensure Universal Uniqueness
+
+**Description:** Modify UUID generation to use a combination of VM name and namespace, ensuring unique UUIDs within the cluster.
```

Comment:
> what about universal uniqueness? How would it be ensured?

Using both name + namespace as a hash will ensure uniqueness AFAICT

Reply:
not universal uniqueness. a different vm on a different cluster would obtain the same uuid if it has the same name and namespace.

Reply:
Ah, got you.
We can always add something "random" to the hash, like creation timestamp.

Reply:
Would it be possible to clarify the impact of VMs in different clusters having the same UUID if they share the same name and namespace?
Are there specific cases where cross-cluster uniqueness is essential for the UUID to function correctly?

Reply:
I guess that @dankenigsberg is referring migrating a VM from a different cluster, e.g. via importing/exporting disks

Reply:
> Are there specific cases where cross-cluster uniqueness is essential for the UUID to function correctly?

By definition, [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)  has to be universally-unique. A lot of services would break if two different VMs present matching identifiers.

Reply:
Just use the pod medata.UID?
https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids

Reply:
you mean the vm.meta.uid?